mobile/examples/phi-3/ios/LocalLLM/LocalLLM/GenAIGenerator.mm

```diff
@@ -5,45 +5,109 @@
 #include "LocalLLM-Swift.h"
 #include "ort_genai.h"
 #include "ort_genai_c.h"
-
+#include <chrono>
 
 @implementation GenAIGenerator
 
-+ (void)generate:(nonnull NSString*)input_user_question {
-  NSString* llmPath = [[NSBundle mainBundle] resourcePath];
-  const char* modelPath = llmPath.cString;
-
-  auto model = OgaModel::Create(modelPath);
-  auto tokenizer = OgaTokenizer::Create(*model);
-
-  NSString* promptString = [NSString stringWithFormat:@"<|user|>\n%@<|end|>\n<|assistant|>", input_user_question];
-  const char* prompt = [promptString UTF8String];
-
-  auto sequences = OgaSequences::Create();
-  tokenizer->Encode(prompt, *sequences);
+typedef std::chrono::high_resolution_clock Clock;
+typedef std::chrono::time_point<Clock> TimePoint;
 
-  auto params = OgaGeneratorParams::Create(*model);
-  params->SetSearchOption("max_length", 200);
-  params->SetInputSequences(*sequences);
-
-  // Streaming Output to generate token by token
-  auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
-
-  auto generator = OgaGenerator::Create(*model, *params);
++ (void)generate:(nonnull NSString*)input_user_question {
+    NSLog(@"Starting token generation...");
+    
+    NSString* llmPath = [[NSBundle mainBundle] resourcePath];
+    const char* modelPath = llmPath.cString;
+    
+    // Log model creation
+    NSLog(@"Creating model ...");
+    auto model = OgaModel::Create(modelPath);
+    if (!model) {
+        NSLog(@"Failed to create model.");
+        return;
+    }
+    
+    NSLog(@"Creating tokenizer...");
+    auto tokenizer = OgaTokenizer::Create(*model);
+    if (!tokenizer) {
+        NSLog(@"Failed to create tokenizer.");
+        return;
+    }
+    
+    auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);
+    
+    // Construct the prompt
+    NSString* promptString = [NSString stringWithFormat:@"<|user|>\n%@<|end|>\n<|assistant|>", input_user_question];
+    const char* prompt = [promptString UTF8String];
+    
+    NSLog(@"Encoding prompt...");
+    auto sequences = OgaSequences::Create();
+    tokenizer->Encode(prompt, *sequences);
+    
+    // Log parameters
+    NSLog(@"Setting generator parameters...");
+    auto params = OgaGeneratorParams::Create(*model);
+    params->SetSearchOption("max_length", 200);
+    params->SetInputSequences(*sequences);
+    
+    NSLog(@"Creating generator...");
+    auto generator = OgaGenerator::Create(*model, *params);
+    
+    bool isFirstToken = true;
+    TimePoint startTime = Clock::now();
+    TimePoint firstTokenTime;
+    int tokenCount = 0;
+    
+    NSLog(@"Starting token generation loop...");
+    while (!generator->IsDone()) {
+        generator->ComputeLogits();
+        generator->GenerateNextToken();
+        
+        if (isFirstToken) {
+            NSLog(@"First token generated.");
+            firstTokenTime = Clock::now();
```

Comment:
nit: logging takes time so we want to record firstTokenTime prior to logging. 

might also be interesting to include per-token timing in the log as that would help get a picture of performance throughout the generation phase. e.g. is the time per token consistent? if not is the variability random or does it gradually increase/decrease? that can provide hints as to potential causes of performance issues (if there are any).

if we do that we might want to accumulate per-token times in a list and log at the end, as inserting log calls for every token inside the loop could significantly affect overall time taken (esp. is the log call is synchronous which apparently NSLog is).

Reply:
saving timings in array and logging them after loop is over